src/modules/dashboard/hooks/useContractABIParser.ts

```diff
@@ -0,0 +1,93 @@
+import { useEffect, useState } from 'react';
+import { AbiItem, keccak256 } from 'web3-utils';
+
+export interface AbiItemExtended extends AbiItem {
+  name: string;
+  type: 'function';
+  action: string;
+  methodSignature: string;
+  signatureHash: string;
+}
+
+export const fetchContractABI = async (contractAddress: string) => {
+  if (!contractAddress) {
+    return [];
+  }
+
+  try {
+    const apiUri = `https://api.etherscan.io/api?module=contract&action=getAbi&address=${contractAddress}&apiKey=${process.env.ETHERSCAN_API_KEY}`;
```

Comment:
This will only work for contracts on mainnet, presumably? If I can add a safe on Aurora, I should be able to interact with contracts on Aurora as well? Should there be compatibility for using other apis like https://aurorascan.dev/apis?

Reply:
You could look at how Gnosis Safe do it.

The TextField - https://github.com/safe-global/safe-react/blob/c413702a0ad5887b356f6bc6f30aabcfcba1882d/src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/ContractABI/index.tsx###

As well as processing the ABI to get the methods - here https://github.com/safe-global/safe-react/blob/c413702a0ad5887b356f6bc6f30aabcfcba1882d/src/logic/contractInteraction/sources/ABIService/index.ts

Reply:
@arrenv @willm30 In conclusion, we are going to need 4 API keys in order to cover all the current networks:
- Polygonscan
- BSCSCAN (Binance)
- Gnosis BlockScout
- Etherscan (the rest of networks)

Reply:
@ArmandoGraterol That seems like a lot to deal with, is that what Gnosis Safe are doing?

Reply:
@arrenv Gnosis Safe is not clear on what it does with most of the networks, at least it's not clear looking at the repo we are taking as an example as only the Rinkeby, Polygon, and Ether configs are there. Those 3 we know can share the etherscan API.

I recently tested making requests with the Gnosis, BSC, and Polygon WITHOUT API keys and:
- It seems like BlockScout allows not having an API key. But, according to the docs: 
`If My Account is enabled, users can add up to 3 API keys to ensure 10 request/second limits. When API Keys are not used, the global default is 50 requests/second.`

- BSC and Polygon also allow requests without API key but they come with this little message:
`K-Missing/Invalid API Key, rate limit of 1/5sec applied`

Re-tested all of them, just to be sure, with the etherscan API key: 
- Polygon and BSC outright tell you that the keys are invalid.
- BlockScout doesn't tell you if the API keys are wrong or right tbh, you can write anything in the API key param and it still gonna deliver the ABI, so I'm assuming they are only going to complain if we reach the limit I mentioned.

What do you think?

Reply:
@ArmandoGraterol I have not looked into it fully either. Just seems a little odd that you would need an API to be able to read the ABIs.

Reply:
@ArmandoGraterol The following would suggest that Gnosis Safe just use the Etherscan api key: https://github.com/safe-global/safe-react/blob/c413702a0ad5887b356f6bc6f30aabcfcba1882d/src/config/index.ts#L137

Reply:
Yeah, it only uses the etherscan API key. So with the comment I left above I was more or less asking if you guys thought that the limit is not a problem for us, because if that's the case then yeah we can survive with just the etherscan key